### ISSUE TITLE
Mac (Designed for iPad) app can't load type FreemiumKit

### ISSUE THREAD

**Issue by emilismayilov095** [FlineDev/FreemiumKit#8]
Hi, on iPad, iPhone no have problems, but on macOS I get this error: 
`SwiftUICore/EnvironmentObject.swift:92: Fatal error: No ObservableObject of type FreemiumKit found. A View.environmentObject(_:) for FreemiumKit may be missing as an ancestor of this view.`
when Paywall showed

**Comment by emilismayilov095** [comment]
I think you need add macCatalyst(.v15) example: 

let package = Package(
    name: "FreemiumKit",
    defaultLocalization: "en",
    platforms: [
        .iOS(.v15), 
        .macOS(.v12), 
        .macCatalyst(.v15), 
        .tvOS(.v15), 
        .visionOS(.v1)
    ],
    products: [.library(name: "FreemiumKit", targets: ["FreemiumKit"])],
    targets: [
        .target(name: "FreemiumKit", dependencies: ["FreemiumKitSDK"], resources: [.process("PrivacyInfo.xcprivacy")]),
        .binaryTarget(name: "FreemiumKitSDK", path: "FreemiumKitSDK.xcframework.zip"),
    ]
)

**Comment by emilismayilov095** [comment]
@Jeehut, can you check please?

**Comment by Jeehut** [comment]
> Hi, on iPad, iPhone no have problems, but on macOS I get this error:
SwiftUICore/EnvironmentObject.swift:92: Fatal error: No ObservableObject of type FreemiumKit found. A View.environmentObject(_:) for FreemiumKit may be missing as an ancestor of this view.
when Paywall showed

This error indicates that you have not added the modifier `.environmentObject(FreemiumKit.shared)` to your root view. Please make sure you follow the [SDK Setup Guide](https://freemiumkit.app/documentation/freemiumkit/setupguide) and specify the modifier to your root view. If everything works on iPhone but not on Mac, then you may have added the modifier to the root view opened on iPhone but not the root view on Mac.

If you believe that you have properly set up the the modifier, please share the code where you do this so I can help.

> I think you need add macCatalyst(.v15) example:
> 
> ```swift
> let package = Package(
>     name: "FreemiumKit",
>     defaultLocalization: "en",
>     platforms: [
>         .iOS(.v15), 
>         .macOS(.v12), 
>         .macCatalyst(.v15), 
>         .tvOS(.v15), 
>         .visionOS(.v1)
>     ],
>     products: [.library(name: "FreemiumKit", targets: ["FreemiumKit"])],
>     targets: [
>         .target(name: "FreemiumKit", dependencies: ["FreemiumKitSDK"], resources: [.process("PrivacyInfo.xcprivacy")]),
>         .binaryTarget(name: "FreemiumKitSDK", path: "FreemiumKitSDK.xcframework.zip"),
>     ]
> )
> ```

If you can successfully build and run the application which you can (even though it crashes), it means that there's no need to add any new platform support to solve your problem. In fact, the destinatino "Mac (Designed for iPad)" is using the iOS platform. Mac Catalyst is a completely separate platform from that. That doesn't seem to be your problem.

**Comment by emilismayilov095** [comment]
@Jeehut  Hi, No, I added modifier .environmentObject(FreemiumKit.shared) and configured everything correctly, since in iOS and iPadOS everything works fine. The problem is only on Mac

**Comment by Jeehut** [comment]
I cannot reproduce your issue here on my end. I have created a demo project and added a "Mac (Designed for iPad)" destination and everything works as expected, I'm not running into your error. I can only see your error, if I remove the environmentObject modifier. Could you send me a minimal reproducible version of your project, either uploaded here as a ZIP file or more private via email to freemiumkit@fline.dev – then I'll gladly investigate.

**Comment by emilismayilov095** [comment]
<img width="2010" alt="Снимок экрана 2024-12-07 в 18 10 17" src="https://github.com/user-attachments/assets/7c46c79a-f070-451e-b858-4a984aedc6a4">

**Comment by Jeehut** [comment]
@emilismayilov095 This is a completely different error message! The error message you posted originally was indicating that you had no `environmentObjectt`, but this one is different:

It clearly states that you haven't added the asset 'FreemiumKit' to the Asset Catalog yet. As described in the [Setup Guide](https://freemiumkit.app/documentation/freemiumkit/setupguide), you need to drag & drop the config file from the 'Setup' tab in the FreemiumKit Mac app to your Asset Catalog. I have downloaded your sample project, dragged a config into the Asset Catalog and it built just fine!

Please try again by following the instructions closely. Maybe double-check if your asset catalog is properly linked to your target. I'm closing this issue since I believe it's just a configuration error. Let me know if you run into any issues and I'll reopen if there's anything wrong with the SDK or the app.

Out of curiosity: Have you read the setup guide and followed that or have you watched the video and followed that? Or did you try figuring things out just by using the app itself? I'd like to understand which part of my Onboarding might need some improvement.

**Comment by emilismayilov095** [comment]
@Jeehut Oh, sorry this is my bad. for reproduce you nee call show paywall.
<img width="2018" alt="Снимок экрана 2024-12-07 в 18 57 08" src="https://github.com/user-attachments/assets/0013dd90-3951-4d8b-aede-360cc1ae9bcc">
[FreemiumKitMacExample.zip](https://github.com/user-attachments/files/18048821/FreemiumKitMacExample.zip)

**Comment by emilismayilov095** [comment]
<img width="1056" alt="Снимок экрана 2024-12-07 в 18 59 02" src="https://github.com/user-attachments/assets/9dea99d0-c098-4ba6-bcae-127524cffeff">

Please check project again.

**Comment by Jeehut** [comment]
<img width="1393" alt="Screenshot 2024-12-07 at 16 14 44" src="https://github.com/user-attachments/assets/7172f80e-1a87-46cb-aeb0-20fc5738a190">
I just downloaded your new project, unzipped, opened it, selected the "My Mac (Designed for iPad)" scheme and hit Cmd + B to build – and the build succeeded! See attached screenshot above and the top bar.

Which version of Xcode are you using? And which macOS version are you currently on? Have you tried cleaning your cache and clean-building?

**Comment by emilismayilov095** [comment]
@Jeehut  please run the project so that when it asks for Paywall it crashes.
<img width="1117" alt="Снимок экрана 2024-12-07 в 19 19 37" src="https://github.com/user-attachments/assets/135a5c29-2650-4b4e-8684-99d9497cad5f">

Sequoia 15.1.1

**Comment by Jeehut** [comment]
@emilismayilov095 Ah, my bad this time! 🙈

Yeah, when running your project, I do run into the same error. Finally I can reproduce it! 💪
But I won't be able to get back to this before next week. Please stay tuned. Reopening…

**Comment by emilismayilov095** [comment]
@Jeehut  thank you. Waiting you. I really love Freemium Kit, and this is my second project with this library.

**Comment by Jeehut** [comment]
@emilismayilov095 Thank you for the kind words! Glad to hear that.

I did have time to investigate real quick and I can already say that it looks like it's a SwiftUI bug. In particular, what I'm doing is the `.paywall(isPresented:)` modifier is adding a sheet to your view hierarchy which shows a `PaywallView`. But for some reason, when running through "Mac (Designed for iPad)", that view inside the sheet doesn't get access to the environment object. I strongly believe that this is a bug in SwiftUI. I will make sure to report to Apple and attempt to find a workaround so you're not stuck. Just wanted to let you know my immediate findings.

**Comment by emilismayilov095** [comment]
@Jeehut  Hi, any news Bro?

**Comment by Jeehut** [comment]
I've tried 2 ideas for a workaround, but they didn't work. Will try a few more approaches later this week, I'll let you know when I found a working one. The weird thing is, even if you put an .environmentObject(FreemiumKit.shared) right below the .paywall call, SwiftUI still doesn't pass it along properly!

It's definitely a bug in SwiftUI, and I will report a bug to Apple once I tried all approaches.

**Comment by Jeehut** [comment]
@emilismayilov095 I have now found a workaround and released [version `1.16.1`](https://github.com/FlineDev/FreemiumKit/releases/tag/1.16.1) that should fix this crash. I am also reporting a bug to Apple next so they can fix it and I can get rid of the workaround sometime. Let me know if you still run into any issues!

**Comment by Jeehut** [comment]
For reference, here's a minimal code example that crashes when run on "My Mac (Designed for iPad)" in macOS 15.2, proving that this is indeed a SwiftUI bug that Apple needs to fix (the same code works just fine in iOS):

```swift
import SwiftUI

struct ContentView: View {
   @State private var myObject: MyObject = .init()
   @State private var showCustomSheet: Bool = false

   var body: some View {
      VStack {
         // some view
      }
      .fullScreenCover(isPresented: self.$showCustomSheet) {
         CustomSheetView()
      }
      .environmentObject(self.myObject)
      .onAppear {
         self.showCustomSheet = true
      }
   }
}

final class MyObject: ObservableObject {
   @Published
   var value: String = "Hello, World!"
}

struct CustomSheetView: View {
   @EnvironmentObject private var myObject: MyObject

   var body: some View {
      Text(self.myObject.value)
   }
}
```

**Comment by emilismayilov095** [comment]
@Jeehut  Hello, you are amazing, work Perfect!! Thank you!

**Comment by Jeehut** [comment]
Posting a clean summary here for future readers — including AI coding agents whose users hit this — so the workaround is easy to find without scrolling through the back-and-forth above. The fix shipped in [v1.16.1](https://github.com/FlineDev/FreemiumKit/releases/tag/1.16.1).

## The bug

When an iOS app runs on Mac via "My Mac (Designed for iPad)" (i.e. `ProcessInfo.processInfo.isiOSAppOnMac == true`), SwiftUI fails to propagate `@EnvironmentObject` values through `.sheet` / `.fullScreenCover` modifiers. So when `.paywall(isPresented:)` presented `PaywallView` as a sheet, its `@EnvironmentObject private var freemiumKit: FreemiumKit` came up empty and the app crashed with the classic *"No ObservableObject of type FreemiumKit found"* message.

It only happens on Mac (Designed for iPad) — pure iPad, iPhone, and macOS native builds were fine. Even adding `.environmentObject(FreemiumKit.shared)` directly under `.paywall(...)` did not help. I filed this as a SwiftUI bug with Apple under **FB16112395**. A minimal repro (a `.fullScreenCover` with an `@EnvironmentObject`) is earlier in this thread.

## The workaround

In `PaywallView`, the `@EnvironmentObject` was renamed to a backing property and wrapped behind a computed property that bypasses the environment on Mac (Designed for iPad) by going straight to the `FreemiumKit.shared` singleton:

```swift
struct PaywallView: View {
   @EnvironmentObject private var freemiumKitFromEnvironment: FreemiumKit

   // NOTE: workaround for a SwiftUI bug, see issue #8
   private var freemiumKit: FreemiumKit {
      #if os(iOS)
      if ProcessInfo.processInfo.isiOSAppOnMac {
         FreemiumKit.shared
      } else {
         self.freemiumKitFromEnvironment
      }
      #else
      self.freemiumKitFromEnvironment
      #endif
   }
   // ... rest of the view unchanged ...
}
```

It works because `FreemiumKit` is a singleton — on Mac (Designed for iPad) we skip the broken environment lookup and read `FreemiumKit.shared` directly. Everywhere else (iOS proper, real macOS, visionOS, tvOS, watchOS) it keeps using the normal `@EnvironmentObject`, so there are no side effects on platforms where SwiftUI behaves correctly. Internal call sites are unchanged because the new computed property kept the original name `freemiumKit`.

If you hit the same crash in **your own** view (not `PaywallView`) when presenting it via `.sheet` / `.fullScreenCover` from a "Mac (Designed for iPad)" target, the same pattern applies: keep the `@EnvironmentObject` as a fallback, but on `isiOSAppOnMac` read `FreemiumKit.shared` directly.
